In [16]:
# Import Libraries 
import numpy as np
import pandas as pd

In [17]:
import os

df = pd.read_csv(os.path.expanduser("~/Documents/hospital_intelligence/data/raw/healthcare_dataset.csv"))

In [18]:
# fetch first 5 rows
df.head()

,Name,Age,Gender,Blood Type,Medical Condition,Date of Admission,Doctor,Hospital,Insurance Provider,Billing Amount,Room Number,Admission Type,Discharge Date,Medication,Test Results
0,Bobby JacksOn,30,Male,B-,Cancer,2024-01-31,Matthew Smith,Sons and Miller,Blue Cross,18856.281306,328,Urgent,2024-02-02,Paracetamol,Normal
1,LesLie TErRy,62,Male,A+,Obesity,2019-08-20,Samantha Davies,Kim Inc,Medicare,33643.327287,265,Emergency,2019-08-26,Ibuprofen,Inconclusive
2,DaNnY sMitH,76,Female,A-,Obesity,2022-09-22,Tiffany Mitchell,Cook PLC,Aetna,27955.096079,205,Emergency,2022-10-07,Aspirin,Normal
3,andrEw waTtS,28,Female,O+,Diabetes,2020-11-18,Kevin Wells,"Hernandez Rogers and Vang,",Medicare,37909.782410,450,Elective,2020-12-18,Ibuprofen,Abnormal
4,adrIENNE bEll,43,Female,AB+,Cancer,2022-09-19,Kathleen Hanna,White-White,Aetna,14238.317814,458,Urgent,2022-10-09,Penicillin,Abnormal


In [19]:
# fetch bottom 5 records
df.tail()

,Name,Age,Gender,Blood Type,Medical Condition,Date of Admission,Doctor,Hospital,Insurance Provider,Billing Amount,Room Number,Admission Type,Discharge Date,Medication,Test Results
55495,eLIZABeTH jaCkSOn,42,Female,O+,Asthma,2020-08-16,Joshua Jarvis,Jones-Thompson,Blue Cross,2650.714952,417,Elective,2020-09-15,Penicillin,Abnormal
55496,KYle pEREz,61,Female,AB-,Obesity,2020-01-23,Taylor Sullivan,Tucker-Moyer,Cigna,31457.797307,316,Elective,2020-02-01,Aspirin,Normal
55497,HEATher WaNG,38,Female,B+,Hypertension,2020-07-13,Joe Jacobs DVM,"and Mahoney Johnson Vasquez,",UnitedHealthcare,27620.764717,347,Urgent,2020-08-10,Ibuprofen,Abnormal
55498,JENniFER JOneS,43,Male,O-,Arthritis,2019-05-25,Kimberly Curry,"Jackson Todd and Castro,",Medicare,32451.092358,321,Elective,2019-05-31,Ibuprofen,Abnormal
55499,jAMES GARCiA,53,Female,O+,Arthritis,2024-04-02,Dennis Warren,Henry Sons and,Aetna,4010.134172,448,Urgent,2024-04-29,Ibuprofen,Abnormal


In [20]:
# statistical description 
df.describe()

,Age,Billing Amount,Room Number
count,55500.000000,55500.000000,55500.000000
mean,51.539459,25539.316097,301.134829
std,19.602454,14211.454431,115.243069
min,13.000000,-2008.492140,101.000000
25%,35.000000,13241.224652,202.000000
50%,52.000000,25538.069376,302.000000
75%,68.000000,37820.508436,401.000000
max,89.000000,52764.276736,500.000000


In [21]:
# check total null values 
df.isnull().sum()

Name                  0
Age                   0
Gender                0
Blood Type            0
Medical Condition     0
Date of Admission     0
Doctor                0
Hospital              0
Insurance Provider    0
Billing Amount        0
Room Number           0
Admission Type        0
Discharge Date        0
Medication            0
Test Results          0
dtype: int64

In [22]:
# total rows and columns
df.shape

(55500, 15)

In [23]:
# lowest billing amount 
df["Billing Amount"].min()

-2008.4921398591305

In [24]:
# types of each column
df.dtypes

Name                   object
Age                     int64
Gender                 object
Blood Type             object
Medical Condition      object
Date of Admission      object
Doctor                 object
Hospital               object
Insurance Provider     object
Billing Amount        float64
Room Number             int64
Admission Type         object
Discharge Date         object
Medication             object
Test Results           object
dtype: object

In [25]:
# total of each admission type
df["Admission Type"].value_counts()

Admission Type
Elective     18655
Urgent       18576
Emergency    18269
Name: count, dtype: int64

In [26]:
# total of each Medical Condition
df["Medical Condition"].value_counts()

Medical Condition
Arthritis       9308
Diabetes        9304
Hypertension    9245
Obesity         9231
Cancer          9227
Asthma          9185
Name: count, dtype: int64

In [27]:
# fixing the type of date column
df["Date of Admission"] = pd.to_datetime(df["Date of Admission"])
df["Discharge Date"] = pd.to_datetime(df["Discharge Date"])

In [47]:
# Adding a new column 
df["length_of_stay"] = (df["discharge_date"] - df["date_of_admission"]).dt.days

In [48]:
df.columns

Index(['patient_id', 'name', 'age', 'gender', 'blood_type',
       'medical_condition', 'date_of_admission', 'doctor', 'hospital',
       'insurance_provider', 'billing_amount', 'room_number', 'admission_type',
       'discharge_date', 'medication', 'test_results', 'readmission_flag',
       'length_of_stay'],
      dtype='object')

In [49]:
df["length_of_stay"].describe()

count    55392.000000
mean        15.508774
std          8.659552
min          1.000000
25%          8.000000
50%         15.000000
75%         23.000000
max         30.000000
Name: length_of_stay, dtype: float64

In [31]:
# total count records with negative Billing Amount
df[df["Billing Amount"]<0].shape[0] 

108

Since the billing amount cannot be negative, and only 108 records (less than 0.2% of the total dataset) have values below zero, these records are being removed from the dataset.

In [32]:
# drop  records with negative Billing Amount
df.drop(df[df["Billing Amount"]<0].index, inplace = True)

In [33]:
# changing the case of column names and replacing the ' ' with '_' 
df.columns = df.columns.str.lower().str.strip().str.replace(" ","_")

In [34]:
# Correcting name casing in the 'name' column
df['name'] = df['name'].str.title().str.strip().str.replace("_"," ")

In [35]:
# Addinf new column by generating unique id for each 
df["patient_id"] = ["P"+str(i).zfill(5) for i in range(1, len(df)+1)]

In [36]:
cols = df.columns

In [37]:
# place the 'patient_id' in the first place
df = df[['patient_id'] + [col for col in cols if col != 'patient_id']]

In [38]:
# the rate of readmitted patients based on bench mark
df["readmission_flag"] = np.random.choice([0,1], size=len(df), p=[0.85, 0.15])

In [39]:
df["readmission_flag"].value_counts()

readmission_flag
0    47027
1     8365
Name: count, dtype: int64

df.shape

##### Readmission flag
Simulated using 85/15 probability based on the industry benchmark of 15–20%. Actual rate came out to ~14.7% (8,143 out of 55,392 patients), which is within the acceptable range.

#### Data generation & cleaning — conclusion
The original dataset contained 55,500 rows and 15 columns sourced from Kaggle. During inspection, 108 rows with negative billing amounts were identified and removed as they are logically invalid in a hospital context, representing less than 0.2% of total data.

Three new columns were engineered: length_of_stay (derived from admission and discharge dates), patient_id (unique identifier in format P00001–P55392), and readmission_flag (simulated based on industry benchmark).
The industry average hospital readmission rate is 15–20%. Our simulated rate came out to ~14.7%, calculated as readmitted patients divided by total patients. This slight difference is expected with random simulation and falls well within the acceptable benchmark range.

Final dataset: 55,392 rows × 18 columns, cleaned, typed correctly, and ready for database loading.

In [51]:
df.to_csv("/Users/bulbulgupta/Documents/hospital_intelligence/data/processed/cleaned_healthcare.csv", index=False)